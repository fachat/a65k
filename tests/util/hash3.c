

#include <string.h>

#define	DEBUG

#include "log.h"
#include "astring.h"
#include "hashmap.h"

typedef struct {
	const char *name;
	const char *data;
} entry_t;

static entry_t data[] = {
	{ "a", "11" }, 
	{ "b", "22" }, 
	{ "c", "33" }, 
	{ "d", "44" }, 
	{ "e", "55" }, 
	{ "f", "66" }, 
	{ "g", "77" }, 
	{ "h", "88" }, 
	{ "i", "99" },
	{ "nmos_illegal_nobcd", "test has negative hash used to cause illegal pointer ref" },
};

static bool_t equals_key(const void *data1, const void *data2) {

	log_debug("equals for '%s' vs. '%s' is %d", data1, data2, strcmp((char*)data1, (char*)data2));

	return (0 == strcmp((char*)data1, (char*)data2));
}

static const void* key_from_entry(const void *data) {

	entry_t *entry = (entry_t*) data;

	return 	entry->name;
}


// short hash algorithm to provoke hash conflicts
static int shorthash_from_key(const void *data) {
	
	char *cdata = (char*)data;

	int len = strlen(cdata);
	
	int hash = len;

	// the hash algorithm used below only makes sense
	// for up to 2 bytes 
	if (len > 2) {
		len = 2;
	}
	for (int i = 0; i < len; i++) {
		hash = hash * 13 + (cdata[i] & 0x07);
	}

	log_debug("hash for '%s' is %d", cdata, hash);

	return hash;
}


void do_test(hash_t *hash);

int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

        log_module_init(LEV_DEBUG);

	hash_t *hash = NULL;

	log_debug("%s", "hashmap:");

	hash = hash_init(3, 3, shorthash_from_key, key_from_entry, equals_key);

	do_test(hash);

	return 0;
}


void do_test(hash_t *hash) {

	int datasize = sizeof(data)/sizeof(entry_t);
	printf ("sizeof data=%d\n", datasize);
	
	for (int i = 0; i < datasize; i++) {
		
		hash_put(hash, (void*)&(data[i]));
	}

	printf("-------- len =%ld\n", hash_size(hash));
	hash_iterator_t *iter = hash_iterator(hash);
	void *v = NULL;
	while ( (v = hash_iterator_next(iter)) ) {
		printf("k:%s, v:%s\n", ((entry_t*)v)->name, ((entry_t*)v)->data);
	}

	v = hash_iterator_next(iter);
	printf("excess call returns: %p\n", v);

	entry_t e1 = { "foo", "bar" };
	hash_put(hash, (void*) &e1);

	printf("-------- len =%ld\n", hash_size(hash));
	iter = hash_iterator(hash);
	while ( (v = hash_iterator_next(iter)) ) {
		printf("k:%s, v:%s\n", ((entry_t*)v)->name, ((entry_t*)v)->data);
	}

	printf("-------- len =%ld\n", hash_size(hash));
	entry_t e2 = { "foo2", "bar" };
	int i = 0;
	iter = hash_iterator(hash);
	while ( (v = hash_iterator_next(iter)) ) {
		if (i++ == 4) {
			hash_put(hash, &e2);
		}
		printf("k:%s, v:%s\n", ((entry_t*)v)->name, ((entry_t*)v)->data);
	}
}



