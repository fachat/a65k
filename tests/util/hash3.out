DBG:hashmap:
sizeof data=10
DBG:hash for 'a' is 14
DBG:hash for 'b' is 15
DBG:hash for 'c' is 16
DBG:hash for 'd' is 17
DBG:hash for 'e' is 18
DBG:hash for 'f' is 19
DBG:hash for 'g' is 20
DBG:hash for 'h' is 13
DBG:hash for 'i' is 14
DBG:equals for 'a' vs. 'i' is -8
DBG:hash for 'nmos_illegal_nobcd' is 3125
-------- len =10
k:b, v:22
k:e, v:55
k:c, v:33
k:f, v:66
k:h, v:88
k:a, v:11
k:d, v:44
k:g, v:77
k:i, v:99
k:nmos_illegal_nobcd, v:test has negative hash used to cause illegal pointer ref
excess call returns: (nil)
DBG:hash for 'foo' is 592
-------- len =11
k:b, v:22
k:e, v:55
k:c, v:33
k:f, v:66
k:h, v:88
k:foo, v:bar
k:a, v:11
k:d, v:44
k:g, v:77
k:i, v:99
k:nmos_illegal_nobcd, v:test has negative hash used to cause illegal pointer ref
-------- len =11
k:b, v:22
k:e, v:55
k:c, v:33
k:f, v:66
DBG:hash for 'foo2' is 761
k:h, v:88
ERR:Hash modification count mismatch! expected 11, was 12

k:foo, v:bar
ERR:Hash modification count mismatch! expected 11, was 12

k:a, v:11
ERR:Hash modification count mismatch! expected 11, was 12

k:d, v:44
ERR:Hash modification count mismatch! expected 11, was 12

k:g, v:77
ERR:Hash modification count mismatch! expected 11, was 12

k:i, v:99
ERR:Hash modification count mismatch! expected 11, was 12

k:nmos_illegal_nobcd, v:test has negative hash used to cause illegal pointer ref
ERR:Hash modification count mismatch! expected 11, was 12

k:foo2, v:bar
ERR:Hash modification count mismatch! expected 11, was 12

